### PR TITLE
OTP-1382 GMAP Push notifications for leg transition

### PR DESCRIPTION
**Pull request by br648**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

Check for upcoming leg transitions (arrive, departure & mode change) and notify observers.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/i18n/Message.java [hunk 1]
```diff
@@ -19,7 +19,10 @@ public enum Message {
     ACCEPT_DEPENDENT_EMAIL_SUBJECT,
     ACCEPT_DEPENDENT_EMAIL_MANAGE,
     ACCEPT_DEPENDENT_ERROR,
+    DESTINATION_CHANGE_NOTIFICATION,
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/i18n/Message.java]: Rename to something like `ARRIVED_NOTIFICATION` or `AT_DESTINATION_NOTIFICATION`?

File: src/main/java/org/opentripplanner/middleware/i18n/Message.java [hunk 2]
```diff
@@ -19,7 +19,10 @@ public enum Message {
...
     LABEL_AND_CONTENT,
+    MODE_CHANGE_NOTIFICATION,
+    ORIGIN_CHANGE_NOTIFICATION,
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/i18n/Message.java]: Rename to something like `DEPARTED_NOTIFICATION` or `LEAVE_ORIGIN_NOTIFICATION`?

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorNotification.java [hunk 3]
```diff
@@ -110,4 +111,46 @@ public static TripMonitorNotification createInitialReminderNotification(
...
+                changedLeg.mode,
+                expectedLeg.from.name
+            )
+        );
+    }
+
+    /**
+     * Creates a notification that the origin or destination has changed.
+     */
+    public static TripMonitorNotification createStopChangeNotification(
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/TripMonitorNotification.java]: Could you clarify what you had in mind for these notification types? I thought we would need a notification that the person traveling has left their origin or arrived at the destination.

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java [hunk 4]
```diff
@@ -498,6 +549,95 @@ public TripMonitorNotification checkTripForDelay(NotificationType delayType) {
         return null;
     }
 
+    /**
+     * Check for major upcoming change and notify the traveler.
+     */
+    @Nullable
+    public TripMonitorNotification checkTripForMajorChange(
+        NotificationType delayType,
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Rename this argument to `notificationType`.

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java [hunk 5]
```diff
@@ -498,6 +549,95 @@ public TripMonitorNotification checkTripForDelay(NotificationType delayType) {
...
+     */
+    private List<NotificationType> getAllMajorChangeNotificationTypes() {
+        return List.of(NotificationType.MODE_CHANGE, NotificationType.ORIGIN_CHANGE, NotificationType.DESTINATION_CHANGE);
+    }
+
+    /**
+     * Attempt to match on just a single exception, else use the wider exception criteria.
+     */
+    @Nullable
+    private Itinerary getMatchingItinerary(NotificationType delayType, List<Itinerary> itineraries) {
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Rename `delayType` to `notificationType`

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java [hunk 6]
```diff
@@ -498,6 +549,95 @@ public TripMonitorNotification checkTripForDelay(NotificationType delayType) {
...
+            return !ItineraryUtils.stopsMatch(trip.itinerary.legs.get(index).to, itineraryToCheck.legs.get(index).to);
+        }
+        return false;
+    }
+
+    /**
+     * A list of all notification types that are classed as major changes to an itinerary.
+     */
+    private List<NotificationType> getAllMajorChangeNotificationTypes() {
+        return List.of(NotificationType.MODE_CHANGE, NotificationType.ORIGIN_CHANGE, NotificationType.DESTINATION_CHANGE);
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Could this be a constant?

**Review by binh-dam-ibigroup — COMMENTED** [review]
Could you clarify the intent behind origin/destination change? I think what we want to do is to notify observers that someone has departed or arrived on their trip.

**Comment by br648** [comment]
> Could you clarify the intent behind origin/destination change? I think what we want to do is to notify observers that someone has departed or arrived on their trip.

Oh, really! I thought I was to check the OTP response for any changes to the traveler's trip and notify them of this. Could I have a couple of use cases just to clarify what is expected please?

**Comment by binh-dam-ibigroup** [comment]
> Oh, really! I thought I was to check the OTP response for any changes to the traveler's trip and notify them of this. Could I have a couple of use cases just to clarify what is expected please?

It is simpler than you think it is. Basically, a notification when the traveler leaves the origin or arrives at destination, and when the traveler changes modes (arrives at a bus stop, gets off the bus, gets on/off the bike if there is a bike involved).

**Comment by binh-dam-ibigroup** [comment]
> Basically, a notification when the traveler leaves the origin

Notifications are for observers who are not traveling with the traveler and their companion.

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java [hunk 7]
```diff
@@ -86,6 +91,12 @@ public static String getInstruction(
                 }
             }
         }
+
+        try {
+            new CheckMonitoredTrip(travelerPosition.trip).checkForMajorTripChanges(travelerPosition.nextLeg);
+        } catch (Exception e) {
+            LOG.error("Error encountered while checking for major trip changes.", e);
+        }
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: When a traveler approaches a bus leg from a walk leg, this block doesn't get called. In the case of a walk->bus transition, the block at line 67 is executed instead. The code in this block needs to execute with the logic when arriving at the end of a leg where the next leg is on a different mode.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Could you make these changes please:
- convert the departure/arrival change notification into leaving the origin/arriving at destination notifications, and send the notification when the traveler gets to the corresponding location.
- execute mode change notification when approaching the end of a leg where the next leg has a different mode.

File: src/main/java/org/opentripplanner/middleware/i18n/Message.java (hunk 1 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/i18n/Message.java]: Fixed in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/i18n/Message.java (hunk 2 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/i18n/Message.java]: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorNotification.java (hunk 3 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/models/TripMonitorNotification.java]: Refactored and addressed in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 4 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Refactored and addressed in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 5 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Refactored and addressed in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 6 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Refactored and addressed in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java (hunk 7 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: Refactored and addressed in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /**
+     * Get the traveler's name if available, if not provide a generic traveler name.
+     */
+    private String getTravelerName() {
+        if (travelerName != null) {
+            return travelerName;
+        } else {
+            return Message.TRIP_TRAVELER_GENERIC_NAME.get(observerLocale);
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: Fallback on email if the name is not available (extract/reuse that code), similar to with notification emails about companion requests and shared trips.

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java [hunk 9]
```diff
@@ -247,6 +250,31 @@ private boolean isTrackingOngoing() {
...
+        OtpUser otpUser = getOtpUser();
+        if (otpUser != null) {
+            otpUser.relatedUsers.forEach(relatedUser -> {
+                if (relatedUser.isConfirmed()) {
+                    OtpUser observer = Persistence.otpUsers.getOneFiltered(eq("email", relatedUser.email));
+                    if (observer != null) {
+                        enqueueNotification(
+                            LegTransitionNotification.createLegTransitionNotifications(
+                                legTransitionTypes,
+                                otpUser.name,
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Fallback on the user email (UI is not populating names at this moment.)

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java [hunk 10]
```diff
@@ -525,9 +560,12 @@ private void sendNotifications() {
             return;
         }
 
+        Locale locale = getOtpUserLocale();
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Replace this call to `getOtpUserLocale` with `I18nUtils.getOtpUserLocale(otpUser)` to use the locale of the desired user. Would it make sense to delete method `getOtpUserLocale` because so many calls are user-specific?

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * Create {@link TripMonitorNotification} for leg transition based on notification type.
+     */
+    @Nullable
+    public TripMonitorNotification createTripMonitorNotification(NotificationType notificationType) {
+        String body;
+        switch (notificationType) {
+            case MODE_CHANGE_NOTIFICATION:
+                body = String.format(
+                    Message.MODE_CHANGE_NOTIFICATION.get(travelerPosition.locale),
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: `observerLocale` should be used in all notifications.

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java [hunk 12]
```diff
@@ -371,4 +371,4 @@ public static Leg getFirstLeg(Itinerary itinerary) {
             .map(legs -> legs.get(0))
             .orElse(null);
     }
-}
+}
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: Revert this change?

File: src/test/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTripTest.java [hunk 13]
```diff
@@ -214,6 +220,65 @@ private static Stream<Arguments> createDelayNotificationTestCases() {
...
+    @MethodSource("createLegTransitionNotificationTestCases")
+    void testLegTransitionNotifications(
+        NotificationType legTransitionType,
+        String travelerName,
+        TravelerPosition travelerPosition,
+        Locale locale,
+        String message
+    ) {
+        TripMonitorNotification[] notification = LegTransitionNotification.createLegTransitionNotifications(
+            List.of(legTransitionType),
```
Comment (binh-dam-ibigroup) [inline src/test/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTripTest.java]: I didn't initially get why you were using a list, but I see what that does on transitions... I still think that only one transition type should be dealt with as a time as a person arrives at the end of a leg and begins the next one.

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+                travelerName,
+                legTransitionType,
+                travelerPosition,
+                observerLocale
+            );
+            if (legTransitionNotification.tripMonitorNotification != null) {
+                tripMonitorNotifications.add(legTransitionNotification.tripMonitorNotification);
+            }
+        }
+        return tripMonitorNotifications.toArray(new TripMonitorNotification[0]);
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: Change (and rename) `legTransitionTypes` to be a single `NotificationType`, so we can remove the for loop and still return the corresponding notification.

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java [hunk 15]
```diff
@@ -247,6 +250,31 @@ private boolean isTrackingOngoing() {
         return trip.journeyState.tripStatus == TripStatus.TRIP_ACTIVE && TripTrackingData.getOngoingTrackedJourney(trip.id) != null;
     }
 
+    /**
+     * Process leg transition notifications.
+     */
+    public void processLegTransition(List<NotificationType> legTransitionTypes, TravelerPosition travelerPosition) {
+        OtpUser otpUser = getOtpUser();
+        if (otpUser != null) {
+            otpUser.relatedUsers.forEach(relatedUser -> {
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: The notifications should be sent to the observers on a trip (not to all companions of a user), and to either the primary traveler or the companion on a trip, if they are not the user who started live tracking.

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java [hunk 16]
```diff
@@ -247,6 +250,31 @@ private boolean isTrackingOngoing() {
...
+                if (relatedUser.isConfirmed()) {
+                    OtpUser observer = Persistence.otpUsers.getOneFiltered(eq("email", relatedUser.email));
+                    if (observer != null) {
+                        enqueueNotification(
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: I think we need to send the notification once, similar to the bus driver notifications. Otherwise, if the traveler stays at an origin or a transition spot for a while, we will be sending the same notification email every few seconds...

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java [hunk 17]
```diff
@@ -91,6 +101,46 @@ public static String getInstruction(
...
+
+    private static List<NotificationType> getLegTransitionTypes(TravelerPosition travelerPosition) {
+        List<NotificationType> notificationTypes = new ArrayList<>();
+        if (isAtStartOfLeg(travelerPosition)) {
+            notificationTypes.add(DEPARTED_NOTIFICATION);
+        }
+        if (isAtEndOfLeg(travelerPosition)) {
+            notificationTypes.add(ARRIVED_NOTIFICATION);
+        }
+        if (hasModeChanged(travelerPosition)) {
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: Can the mode change notification include some of the text of the arrival notification too (so you don't have to make two notifications for that and don't have to pass a list of transition types)?

File: src/test/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTripTest.java [hunk 18]
```diff
@@ -214,6 +220,65 @@ private static Stream<Arguments> createDelayNotificationTestCases() {
...
+        Coordinates expectedLegDestinationCoords = new Coordinates(expectedLeg.to);
+        Leg nextLeg = itinerary.legs.get(2);
+        Coordinates nextLegDepartureCoords = new Coordinates(nextLeg.from);
+        return Stream.of(
+            Arguments.of(
+                NotificationType.MODE_CHANGE_NOTIFICATION,
+                travelerName,
+                new TravelerPosition(expectedLeg, nextLeg, expectedLegDestinationCoords),
+                locale,
+                "Obi-Wan has changed transit from TRAM to WALK."
```
Comment (binh-dam-ibigroup) [inline src/test/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTripTest.java]: Refine the text to say that "[traveler] has arrived at transit stop [stop name]", and omit the specific modes.

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java [hunk 19]
```diff
@@ -91,6 +101,46 @@ public static String getInstruction(
...
+        }
+
+    }
+
+    private static List<NotificationType> getLegTransitionTypes(TravelerPosition travelerPosition) {
+        List<NotificationType> notificationTypes = new ArrayList<>();
+        if (isAtStartOfLeg(travelerPosition)) {
+            notificationTypes.add(DEPARTED_NOTIFICATION);
+        }
+        if (isAtEndOfLeg(travelerPosition)) {
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: If arriving from a bus leg with the bus shape in the middle of the road, and the stop is away from the curb, I think the threshold might not be met. Not sure if we want to use the "upcoming" radius instead of the "immediate" radius.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
I think the logic is getting there. I don't think we need to identify/process a list of notification types (one should be fine). Notifications are for the observers and non-trip-initiators, and there are language tweaks.

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 10 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: I've left `getOtpUserLocale()` in place for cases where the notification is for the user related to the trip i.e. `trip.userId`.

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 16 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: The reason for using `CheckMonitoredTrip` is bc in `sendNotifications(OtpUser otpUser)` there is a check to not send notifications if already sent `if (previousJourneyState.lastNotifications.containsAll(notifications) && !hasInitialReminder)`. Then I noticed that `lastNotifications` is never set! So, I have made an update in this area. The advantage of this approach is that the notification will be sent again if not successful the first time.

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java (hunk 8 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 9 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 10 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java (hunk 11 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/test/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTripTest.java (hunk 13 above)
Comment (br648) [reply src/test/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTripTest.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java (hunk 14 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 15 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java (hunk 16 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java (hunk 17 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/test/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTripTest.java (hunk 18 above)
Comment (br648) [reply src/test/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTripTest.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java (hunk 19 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: Updated in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e

File: src/main/java/org/opentripplanner/middleware/i18n/Message.java [hunk 20]
```diff
@@ -19,10 +19,10 @@ public enum Message {
     ACCEPT_DEPENDENT_EMAIL_SUBJECT,
     ACCEPT_DEPENDENT_EMAIL_MANAGE,
     ACCEPT_DEPENDENT_ERROR,
+    ARRIVED_AND_MODE_CHANGE_NOTIFICATION,
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/i18n/Message.java]: I'd say just keep the old name `MODE_CHANGE_NOTIFICATION`.

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java [hunk 21]
```diff
@@ -66,38 +82,69 @@ public TripMonitorNotification createTripMonitorNotification(NotificationType no
...
+        Set<OtpUser> notifyUsers = new HashSet<>();
+
+        if (trip.ownedByPrimary() && trip.companion != null) {
+            notifyUsers.add(Persistence.otpUsers.getOneFiltered(eq("email", trip.companion.email)));
+        } else if (trip.ownedByCompanion() && trip.primary != null) {
+            notifyUsers.add(Persistence.otpUsers.getById(trip.primary.userId));
         }
+
+        trip.observers.forEach(observer -> {
+            if (observer.isConfirmed()) {
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java:97]: Good thought about using confirmed users only, I forgot to do check that when notifying users when sharing trips!

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java [hunk 22]
```diff
@@ -66,38 +82,69 @@ public TripMonitorNotification createTripMonitorNotification(NotificationType no
...
+                notifyUsers.add(Persistence.otpUsers.getOneFiltered(eq("email", observer.email)));
+            }
+        });
+
+        return notifyUsers;
     }
 
     /**
-     * Create locale specific notifications.
+     * If a traveler is on schedule and on either a walk or transit leg check for possible leg transition notification.
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: I think you meant to say "If a traveler is on the route (not deviated)"...

Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: Insert comma before "check".

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java [hunk 23]
```diff
@@ -66,38 +82,69 @@ public TripMonitorNotification createTripMonitorNotification(NotificationType no
...
+                    LOG.error("Error encountered while checking leg transition.", e);
+                }
             }
         }
-        return tripMonitorNotifications.toArray(new TripMonitorNotification[0]);
     }
-}
+
+    /**
+     * Depending on the traveler's proximity to the start/end of a leg return the appropriate notification type.
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java]: Insert comma before "return".

File: src/test/java/org/opentripplanner/middleware/models/LegTransitionNotificationTest.java [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+            travelerPosition,
+            locale
+        ).tripMonitorNotification;
+        assertNotNull(notification);
+        assertEquals(message, notification.body);
+    }
+
+    private static Stream<Arguments> createLegTransitionNotificationTestCases() throws Exception {
+        String travelerName = "Obi-Wan";
+        Locale locale = Locale.US;
```
Comment (binh-dam-ibigroup) [inline src/test/java/org/opentripplanner/middleware/models/LegTransitionNotificationTest.java]: Maybe remove `locale` as a test parameter (set that in the test itself).

File: src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java [hunk 25]
```diff
@@ -559,6 +592,9 @@ private void sendNotifications() {
         // TODO: better handle below when one of the following fails
         if (successEmail || successPush || successSms) {
             notificationTimestampMillis = DateTimeUtils.currentTimeMillis();
+            // Prevent repeated notifications by saving successfully sent notifications.
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java]: This check is at the wrong spot, I see duplicate notifications being sent and recorded (see screenshot)

image

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Could you address duplicate notifications still being sent, please?

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+            }
+        }
+    }
+
+    /**
+     * Depending on the traveler's proximity to the start/end of a leg, return the appropriate notification type.
+     */
+    private static NotificationType getLegTransitionNotificationType(TravelerPosition travelerPosition) {
+        if (isAtStartOfLeg(travelerPosition)) {
+            return DEPARTED_NOTIFICATION;
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java:133]: Nitpick: Change the condition so that departing notifications are sent at the beginning of the trip only.

I found one case where I set two coordinates, one close to the end of a leg, and the next one close to the beginning of the next leg, and I got two companion notifications, one for arriving at the end of the previous leg, and one for departing the next leg, which is redundant with the previous one and is thus unneeded noise. What do you think?

**Review by binh-dam-ibigroup — APPROVED** [review]
Overall, the changes are working and code looks good. The nice-to-have tweak is not blocking.

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java (hunk 26 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java:133]: I think for consistency we leave it as is. The observer would get a notification that the traveler has arrived but not one to tell them they have departed. The next notification would be the traveler arriving at the end of the next leg. Correct?

File: src/main/resources/Message_fr.properties [hunk 27]
```diff
@@ -4,7 +4,10 @@ ACCEPT_DEPENDENT_EMAIL_LINK_TEXT = Accepter la demande
 ACCEPT_DEPENDENT_EMAIL_SUBJECT = Demande d'accompagnateur
 ACCEPT_DEPENDENT_EMAIL_MANAGE = G�rez vos pr�f�rences
 ACCEPT_DEPENDENT_ERROR = La demande d'accompagnateur n'a pas �t� re�ue.
+ARRIVED_NOTIFICATION = TODO %s.
```
Comment (br648) [inline src/main/resources/Message_fr.properties]: @binh-dam-ibigroup can you provide the french messages please!

File: src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java (hunk 26 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/LegTransitionNotification.java:133]: That's what I observed.

File: src/main/resources/Message_fr.properties (hunk 27 above)
Comment (binh-dam-ibigroup) [reply src/main/resources/Message_fr.properties]: Oui ! https://github.com/ibi-group/otp-middleware/pull/275/commits/36b26afa7d532112d72346b040408576db40a705

**Review by JymDyerIBI — APPROVED** [review]
LGTM
